File: zena_mode/gateway_whatsapp.py
import logging
import asyncio
from flask import Flask, request
from twilio.twiml.messaging_response import MessagingResponse
from async_backend import backend
from config_system import config

logger = logging.getLogger("WhatsAppGateway")

app = Flask(__name__)

# --- Security: Whitelist ---
# Numbers should be in E.164 format: '+1234567890'
ALLOWED_NUMBERS = getattr(config, "whatsapp_whitelist", [])


async def get_zenai_response(query: str) -> str:
    """Helper to get non-streaming response for WhatsApp (MMS/SMS limit)."""
    full_response = ""
    try:
        async with backend:
            async for chunk in backend.send_message_async(query):
                full_response += chunk
        return full_response
    except Exception as e:
        logger.error(f"WhatsApp backend error: {e}")
        return f"❌ Backend Error: {str(e)}"


@app.route("/webhook", methods=["POST"])
def whatsapp_webhook():
    """Handle incoming WhatsApp messages from Twilio."""
    from_number = request.values.get("From", "")
    incoming_msg = request.values.get("Body", "")

    # Check Whitelist
    if from_number not in ALLOWED_NUMBERS:
        logger.warning(f"⛔ Unauthorized WhatsApp from {from_number}")
        resp = MessagingResponse()
        resp.message("⛔ Unauthorized access. Please contact the administrator.")
        return str(resp)

    logger.info(f"📲 WhatsApp from {from_number}: {incoming_msg[:50]}...")

    # Since Flask is sync but backend is async, we run the loop
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    response_text = loop.run_until_complete(get_zenai_response(incoming_msg))
    loop.close()

    # Twilio Response
    resp = MessagingResponse()
    msg = resp.message()
    msg.body(response_text)

    return str(resp)


def run_whatsapp_gateway():
    """Start the Flask server for WhatsApp webhooks."""
    port = getattr(config, "whatsapp_port", 5001)
    # Ensure whitelist is refreshed from config
    global ALLOWED_NUMBERS
    ALLOWED_NUMBERS = getattr(config, "whatsapp_whitelist", [])

    logger.info(f"🚀 WhatsApp Gateway (Webhook) is running on port {port}...")
    logger.info(f"📋 Whitelist: {ALLOWED_NUMBERS}")
    app.run(port=port)


if __name__ == "__main__":
    run_whatsapp_gateway()

File: rust_output/src/gateway_whatsapp.rs
use anyhow::{Result, Context};
use crate::async_backend::{backen};
use crate::config_system::{config};
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static APP: std::sync::LazyLock<Flask> = std::sync::LazyLock::new(|| Default::default());

pub static ALLOWED_NUMBERS: std::sync::LazyLock<getattr> = std::sync::LazyLock::new(|| Default::default());

/// Helper to get non-streaming response for WhatsApp (MMS/SMS limit).
pub async fn get_zenai_response(query: String) -> Result<String> {
    // Helper to get non-streaming response for WhatsApp (MMS/SMS limit).
    let mut full_response = "".to_string();
    // try:
    {
        let _ctx = backend;
        {
            // async for
            while let Some(chunk) = backend.send_message_async(query).next().await {
                full_response += chunk;
            }
        }
        full_response
    }
    // except Exception as e:
}

/// Handle incoming WhatsApp messages from Twilio.
pub fn whatsapp_webhook() -> () {
    // Handle incoming WhatsApp messages from Twilio.
    let mut from_number = request.values.get(&"From".to_string()).cloned().unwrap_or("".to_string());
    let mut incoming_msg = request.values.get(&"Body".to_string()).cloned().unwrap_or("".to_string());
    if !ALLOWED_NUMBERS.contains(&from_number) {
        logger.warning(format!("⛔ Unauthorized WhatsApp from {}", from_number));
        let mut resp = MessagingResponse();
        resp.message("⛔ Unauthorized access. Please contact the administrator.".to_string());
        resp.to_string()
    }
    logger.info(format!("📲 WhatsApp from {}: {}...", from_number, incoming_msg[..50]));
    let mut r#loop = asyncio.new_event_loop();
    asyncio.set_event_loop(r#loop);
    let mut response_text = r#loop.run_until_complete(get_zenai_response(incoming_msg));
    r#loop.close();
    let mut resp = MessagingResponse();
    let mut msg = resp.message();
    msg.body(response_text);
    resp.to_string()
}

/// Start the Flask server for WhatsApp webhooks.
pub fn run_whatsapp_gateway() -> () {
    // Start the Flask server for WhatsApp webhooks.
    let mut port = /* getattr */ 5001;
    // global/nonlocal ALLOWED_NUMBERS
    let mut ALLOWED_NUMBERS = /* getattr */ vec![];
    logger.info(format!("🚀 WhatsApp Gateway (Webhook) is running on port {}...", port));
    logger.info(format!("📋 Whitelist: {}", ALLOWED_NUMBERS));
    app::run(/* port= */ port);
}
